scipy/stats/tests/test_distributions.py

```diff
@@ -3852,6 +3852,55 @@ def test_cdf_ppf_sf_isf_tail(self, x, a, expected):
         isf = stats.dgamma.isf(expected, a)
         assert_allclose(isf, -x, rtol=5e-15)
 
+    @pytest.mark.parametrize("a, ref",
+                             [(1.5, 2.0541199559354117),
+                             (1.3, 1.9357296377121247),
+                             (1.1, 1.7856502333412134)])
+    def test_entropy(self, a, ref):
+        # The reference values were calculated with mpmath:
+        # def entropy_dgamma(a):
+        #    def pdf(x):
+        #        A = mp.one / (mp.mpf(2.) * mp.gamma(a))
+        #        B = mp.fabs(x) ** (a - mp.one)
+        #        C = mp.exp(-mp.fabs(x))
+        #        h = A * B * C
+        #        return h
+        #
+        #    return -mp.quad(lambda t: pdf(t) * mp.log(pdf(t)),
+        #                    [-mp.inf, mp.inf])
+        assert_allclose(stats.dgamma.entropy(a), ref, rtol=1e-14)
+
+    @pytest.mark.parametrize("a, ref",
+                             [(1e-100, -1e+100),
+                             (1e-10, -9999999975.858217),
+                             (1e-5, -99987.37111657023),
+                             (1e4, 6.717222565586032),
+                             (1000000000000000.0, 19.38147391121996),
+                             (1e+100, 117.2413403634669)])
+    def test_entropy_entreme_values(self, a, ref):
+        # The reference values were calculated with mpmath:
+        # from mpmath import mp
+        # mp.dps = 50
+        # def second_dgamma(a):
+        #     a = mp.mpf(a)
+        #     if a < 1e15:
+        #         x_1 = a + mp.log(2) + mp.loggamma(a)
+        #         x_2 = (mp.one - a) * mp.digamma(a)
+        #         h = x_1 + x_2
+        #
+        #     else:
+        #         x_1 = mp.one / mp.mpf(2) * (mp.one + mp.log(a) + mp.log(2 * mp.pi))
+        #         h = mp.log(2) + x_1
```

Comment:
With mpmath the asymptotic expansion is not needed, it maintains precision at high values if everything is *mpmathified*. I think I also messed that up in my [previous comment](https://github.com/scipy/scipy/pull/17873#issuecomment-1417065945) by using `1` and not `mp.one`, sorry about that. If we use the original formula instead, this also tests the correctness of the formula we obtained using the asymptotic expansion. What do you think? 

Using plain mpmath I get a nice smooth curve:


![image](https://user-images.githubusercontent.com/40656107/221424114-daa0f956-62a8-4532-be28-3456f365e2a9.png)

<details>

```
from mpmath import mp
import numpy as np
import matplotlib.pyplot as plt

mp.dps = 50

@np.vectorize
def second_dgamma(a):
    a = mp.mpf(a)
    x_1 = a + mp.log(2) + mp.loggamma(a)
    x_2 = (mp.one - a) * mp.digamma(a)
    h = x_1 + x_2
    return float(h)

x = np.logspace(0, 20, 1000)
plt.semilogx(x, np.array(second_dgamma(x), dtype=np.float64))
plt.show()

Reply:
>this also tests the correctness of the formula we obtained using the asymptotic expansion

I doubt that this is necessary. The asymptotic expansion was derived analytically, so it shouldn't need a test to confirm.

Reply:
> With mpmath the asymptotic expansion is not needed, it maintains precision at high values if everything is _mpmathified_. I think I also messed that up in my [previous comment](https://github.com/scipy/scipy/pull/17873#issuecomment-1417065945) by using `1` and not `mp.one`, sorry about that. If we use the original formula instead, this also tests the correctness of the formula we obtained using the asymptotic expansion. What do you think?

I agree we shouldn't use the asymptotic expansion in the mpmath baseline. 


> I doubt that this is necessary. The asymptotic expansion was derived analytically, so it shouldn't need a test to confirm.

It's still a good sanity check. On principle, I think it's best if the `mpmath` baseline defines the function as simply as possible directly from definitions which can be checked in a standard reference.

Reply:
>With mpmath the asymptotic expansion is not needed, it maintains precision at high values if everything is mpmathified. 

I tried to generate the values without the asymptotic expansion, and it starts to run into problems for very large numbers. If we want to do a sanity check, we could just add a test value for some x where 1e10 < x < 1e15, the reason being that the formula uses the asymptotic expansion for x > 1e10 and the function that generates the reference values uses it for x > 1e15.

Reply:
Ok, done.